File: PiMC.py
# PiMC.py
### Ejecutar. El último argumento indica la cantidad de números aleatorios generados
### time mpiexec  --oversubscribe -np 10 python PiMC.py 100000

# Cargamos la librerías
import numpy as np
import matplotlib.pyplot as plt
import functools 
import sys
import random

# MPI
from mpi4py import MPI


comm = MPI.COMM_WORLD
rank=comm.rank
size=comm.size
n=int(sys.argv[1])

### EL proceso 1 será el especializado en generar números aleatorios 
if rank == 0:
    data_x = np.array([random.random() for x in range(n)]) 
    data_y = np.array([random.random() for x in range(n)])

else:
    data_x = None
    data_y = None
### Aquí usamos el scatter para distribuir los datos
chunk_x= np.zeros(int((n/size)))
chunk_y= np.zeros(int((n/size)))

comm.Scatter(data_x,chunk_x,root=0)
comm.Scatter(data_y,chunk_y,root=0)

### Definimos un contador para almacenar los puntos que caen dentro del círculo unitario
count = 0

for x, y in zip(chunk_x,chunk_y) :
  if x * x + y * y < 1:
    count += 1

print('Procesador '+str(rank)+'. Puntos recibidos ', int((n/size)),'. Puntos dento del círculo: '+ str(count))

### La suma global la recibirá el maestro y la mostrará en pantalla 
global_sum = comm.reduce (count , op = MPI .SUM , root =0) 


if rank == 0:
   pi_est=(global_sum/int(sys.argv[1]))*4
   print('Suma total de puntos dentro del círculo: '+ str(global_sum))
   print('Puntos totales: '+str(sys.argv[1]))
   print('La aproximación de pi es: '+str(pi_est))

File: PiMC.jl
## PiMC.jl

#### Ejecutar. El último argumento indica la cantidad de números aleatorios generados
# time mpirun -np 10 julia PiMC.jl 100000

using MPI
using LinearAlgebra
using Statistics
using Random
using SharedArrays


MPI.Init()
comm = MPI.COMM_WORLD
rank = MPI.Comm_rank(comm)
size = MPI.Comm_size(comm)
n=parse(Int,ARGS[1])
elements=Int(floor(n/size))

puntos_totales=size*elements

root = 0

### Función para calcular los puntos en el circulo
function compute_points(arregloUno,arregloDos,N)
    count=0
    for i=1:N
        r2 = arregloUno[i]^2 + arregloDos[i]^2
        if r2<1.0
            count+=1
        end
    end
    return count

end

if rank == root
   
    #Definimos una semilla
    rng_x = MersenneTwister(1234);
    rng_y = MersenneTwister(5678);
    #Definimos el arreglo de números aleatorios	
    data_x=(rand(rng_x,n))
    data_y=(rand(rng_y,n))
    
else
    data_x = zeros(Float64, 1, elements)
    data_y = zeros(Float64, 1, elements)

end

data_x = MPI.Scatter(data_x,elements,root, comm)
data_y = MPI.Scatter(data_y,elements,root, comm)

### Calculamos los puntos dentro del círculo unitario
n_in_circle=compute_points(data_x,data_y,elements)


### Realizamos el reduce
sr = MPI.Reduce(n_in_circle, +, root, comm)	

println("Procesador $rank . Puntos recibidos: $elements . Puntos dentro del círculo: $n_in_circle")	



if rank == root
   ## Calculamos pi
   pi_est=(sr/puntos_totales)*4.0	
   println(" ")
   println("Suma total de puntos dentro del círculo: $sr")
   println("Puntos totales: $puntos_totales")	
   println("La aproximación de pi es : $pi_est")		
	
end	

MPI.Finalize()





